[Pass Bug] Barriers marked as illegal, conversion failed

### Pass / pipeline name

pto-insert-sync

### Command line

ptoas "$SCRIPT_DIR/${CASE_ID}.pto" -o "$SCRIPT_DIR/${CASE_ID}.cpp"

### Before IR (input)

```markdown
module {
  func.func @float16_n32(%arg0: !pto.ptr<f16>, %arg1: !pto.ptr<f16>, %arg2: i32) {
    %c0 = arith.constant 0 : index
    %c1 = arith.constant 1 : index
    %c32 = arith.constant 32 : index
    %c16 = arith.constant 16 : index
    %c5 = arith.constant 5 : index
    %0 = arith.index_cast %arg2 : i32 to index
    %1 = pto.get_block_idx
    %2 = pto.get_subblock_idx
    %3 = pto.get_subblock_num
    %4 = arith.muli %1, %3 : i64
    %5 = arith.addi %4, %2 : i64
    %6 = pto.get_block_num
    %7 = arith.index_cast %5 : i64 to index
    %8 = arith.index_cast %6 : i64 to index
    %9 = arith.muli %0, %c32 : index
    %10 = arith.ceildivsi %0, %8 : index
    %11 = arith.muli %7, %10 : index
    pto.section.vector {
      %12 = pto.make_tensor_view %arg0, shape = [%9] strides = [%c1] : !pto.tensor_view<?xf16>
      %13 = pto.make_tensor_view %arg1, shape = [%9] strides = [%c1] : !pto.tensor_view<?xf16>
      %14 = pto.alloc_tile : !pto.tile_buf<loc=vec, dtype=f16, rows=1, cols=16, v_row=1, v_col=16, blayout=row_major, slayout=none_box, fractal=512, pad=0>
      %15 = pto.alloc_tile : !pto.tile_buf<loc=vec, dtype=f16, rows=1, cols=16, v_row=1, v_col=16, blayout=row_major, slayout=none_box, fractal=512, pad=0>
      %16 = pto.alloc_tile : !pto.tile_buf<loc=vec, dtype=f16, rows=1, cols=16, v_row=1, v_col=16, blayout=row_major, slayout=none_box, fractal=512, pad=0>
      %17 = pto.alloc_tile : !pto.tile_buf<loc=vec, dtype=f16, rows=1, cols=16, v_row=1, v_col=16, blayout=row_major, slayout=none_box, fractal=512, pad=0>
      %18 = pto.alloc_tile : !pto.tile_buf<loc=vec, dtype=f16, rows=1, cols=32, v_row=1, v_col=32, blayout=row_major, slayout=none_box, fractal=512, pad=0>
      %19 = arith.cmpi slt, %11, %0 : index
      scf.if %19 {
        %20 = arith.addi %11, %10 : index
        %21 = arith.cmpi sgt, %20, %0 : index
        %22 = arith.subi %0, %11 : index
        %23 = arith.select %21, %22, %10 : index
        scf.for %arg3 = %c0 to %23 step %c1 {
          %24 = arith.addi %arg3, %11 : index
          %25 = arith.muli %24, %c32 : index
          %26 = pto.partition_view %12, offsets = [%25], sizes = [%c32] : !pto.tensor_view<?xf16> -> !pto.partition_tensor_view<1x32xf16>
          %27 = arith.addi %25, %c16 : index
          %28 = pto.partition_view %12, offsets = [%27], sizes = [%c16] : !pto.tensor_view<?xf16> -> !pto.partition_tensor_view<1x16xf16>
          %29 = pto.partition_view %13, offsets = [%25], sizes = [%c32] : !pto.tensor_view<?xf16> -> !pto.partition_tensor_view<1x32xf16>
          %30 = pto.partition_view %13, offsets = [%25], sizes = [%c16] : !pto.tensor_view<?xf16> -> !pto.partition_tensor_view<1x16xf16>
          %31 = arith.addi %25, %c16 : index
          %32 = pto.partition_view %13, offsets = [%31], sizes = [%c16] : !pto.tensor_view<?xf16> -> !pto.partition_tensor_view<1x16xf16>
          pto.tload ins(%26 : !pto.partition_tensor_view<1x32xf16>) outs(%18 : !pto.tile_buf<loc=vec, dtype=f16, rows=1, cols=32, v_row=1, v_col=32, blayout=row_major, slayout=none_box, fractal=512, pad=0>)
          pto.barrier_sync[<TLOAD>]
          pto.tload ins(%28 : !pto.partition_tensor_view<1x16xf16>) outs(%15 : !pto.tile_buf<loc=vec, dtype=f16, rows=1, cols=16, v_row=1, v_col=16, blayout=row_major, slayout=none_box, fractal=512, pad=0>)
          pto.barrier_sync[<TLOAD>]
          scf.for %arg4 = %c0 to %c1 step %c1 {
            pto.tgather ins(%18, {maskPattern = #pto.mask_pattern<P0101>} : !pto.tile_buf<loc=vec, dtype=f16, rows=1, cols=32, v_row=1, v_col=32, blayout=row_major, slayout=none_box, fractal=512, pad=0>) outs(%14 : !pto.tile_buf<loc=vec, dtype=f16, rows=1, cols=16, v_row=1, v_col=16, blayout=row_major, slayout=none_box, fractal=512, pad=0>)
            pto.barrier_sync[<TVEC>]
            pto.tadd ins(%14, %15 : !pto.tile_buf<loc=vec, dtype=f16, rows=1, cols=16, v_row=1, v_col=16, blayout=row_major, slayout=none_box, fractal=512, pad=0>, !pto.tile_buf<loc=vec, dtype=f16, rows=1, cols=16, v_row=1, v_col=16, blayout=row_major, slayout=none_box, fractal=512, pad=0>) outs(%16 : !pto.tile_buf<loc=vec, dtype=f16, rows=1, cols=16, v_row=1, v_col=16, blayout=row_major, slayout=none_box, fractal=512, pad=0>)
            pto.barrier_sync[<TVEC>]
            pto.tsub ins(%14, %15 : !pto.tile_buf<loc=vec, dtype=f16, rows=1, cols=16, v_row=1, v_col=16, blayout=row_major, slayout=none_box, fractal=512, pad=0>, !pto.tile_buf<loc=vec, dtype=f16, rows=1, cols=16, v_row=1, v_col=16, blayout=row_major, slayout=none_box, fractal=512, pad=0>) outs(%17 : !pto.tile_buf<loc=vec, dtype=f16, rows=1, cols=16, v_row=1, v_col=16, blayout=row_major, slayout=none_box, fractal=512, pad=0>)
            pto.barrier_sync[<TVEC>]
            pto.tstore ins(%16 : !pto.tile_buf<loc=vec, dtype=f16, rows=1, cols=16, v_row=1, v_col=16, blayout=row_major, slayout=none_box, fractal=512, pad=0>) outs(%30 : !pto.partition_tensor_view<1x16xf16>)
            pto.barrier_sync[<TSTORE_VEC>]
            pto.tstore ins(%17 : !pto.tile_buf<loc=vec, dtype=f16, rows=1, cols=16, v_row=1, v_col=16, blayout=row_major, slayout=none_box, fractal=512, pad=0>) outs(%32 : !pto.partition_tensor_view<1x16xf16>)
            pto.barrier_sync[<TSTORE_VEC>]
            pto.tload ins(%29 : !pto.partition_tensor_view<1x32xf16>) outs(%18 : !pto.tile_buf<loc=vec, dtype=f16, rows=1, cols=32, v_row=1, v_col=32, blayout=row_major, slayout=none_box, fractal=512, pad=0>)
            pto.barrier_sync[<TLOAD>]
            pto.tload ins(%32 : !pto.partition_tensor_view<1x16xf16>) outs(%15 : !pto.tile_buf<loc=vec, dtype=f16, rows=1, cols=16, v_row=1, v_col=16, blayout=row_major, slayout=none_box, fractal=512, pad=0>)
            pto.barrier_sync[<TLOAD>]
          }
        }
      }
    }
    return
  }
}
```

### Expected behavior / expected IR (after)

```markdown
Expected no error and full compilation
```

### Actual IR or error

```text
DEBUG: Start PTOToEmitC Pass
[Debug] Converting MemRef: memref<?xf16>
  [Success] -> __gm__ half*
[Debug] Converting MemRef: memref<1x16xf16, strided<[16, 1]>, #pto.address_space<vec>>
  [Success] -> __ubuf__ half*
[Debug] Converting MemRef: memref<1x32xf16, strided<[32, 1]>, #pto.address_space<vec>>
  [Success] -> __ubuf__ half*
[Debug] Converting MemRef: memref<?xf16, strided<[?], offset: ?>>
  [Success] -> __gm__ half*
[Debug] Converting MemRef: memref<32xf16, strided<[?], offset: ?>>
  [Success] -> __gm__ half*
[Debug] Converting MemRef: memref<1x32xf16, strided<[32, 1], offset: ?>, #pto.address_space<vec>>
  [Success] -> __ubuf__ half*
loc("/mounted_home/pto-dsl/examples/aot/hadamard_multicore/float16_n32.pto":45:11): error: failed to legalize operation 'pto.barrier_sync' that was explicitly marked illegal
Conversion FAILED! Rolling back executed.
Error: Pass execution failed.
Traceback (most recent call last):
  File "/mounted_home/pto-dsl/examples/aot/hadamard_multicore/run_hadamard.py", line 96, in <module>
    run(dtype, n)
  File "/mounted_home/pto-dsl/examples/aot/hadamard_multicore/run_hadamard.py", line 60, in run
    compile_kernel(dtype, n)
  File "/mounted_home/pto-dsl/examples/aot/hadamard_multicore/run_hadamard.py", line 35, in compile_kernel
    subprocess.check_call(
  File "/usr/local/python3.11.14/lib/python3.11/subprocess.py", line 413, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['bash', '/mounted_home/pto-dsl/examples/aot/hadamard_multicore/compile.sh', 'float16', '32']' returned non-zero exit status 1.
[ERROR] 2026-03-03-12:15:43 (PID:3788, Device:6, RankID:-1) ERR99999 UNKNOWN applicaiton exception
```

### Git commit

1e1fbc9098228fa1100572e3a2d44edae1080d3b

### Target Ascend arch (if relevant)

a3

### PTOAS build level (if relevant)

N/A

### Additional context

_No response_

--- a/lib/PTO/Transforms/LoweringSyncToPipe.cpp
+++ b/lib/PTO/Transforms/LoweringSyncToPipe.cpp
@@ -120,13 +120,13 @@ struct BarrierSyncLowering : public OpRewritePattern<BarrierSyncOp> {
   LogicalResult matchAndRewrite(BarrierSyncOp op,
                                 PatternRewriter &rewriter) const override {
     SyncOpType ty = op.getOpType().getOpType();
-    // Only support TMATMUL / TVEC for now
-    if (ty != SyncOpType::TMATMUL && ty != SyncOpType::TVEC)
-      return op.emitError("barrier_sync supports only TMATMUL or TVEC");
-
     PIPE pipe = getPipeFromOpType(ty);
-    if (pipe == PIPE::PIPE_UNASSIGNED)
-      return op.emitError("Failed to map SyncOpType to hardware pipe during barrier lowering.");
+    if (pipe == PIPE::PIPE_UNASSIGNED) {
+      auto diag = op.emitError(
+          "barrier_sync failed to map SyncOpType to hardware pipe during lowering: ");
+      diag << op.getOpType();
+      return failure();
+    }
 
     rewriter.replaceOpWithNewOp<BarrierOp>(
         op, PipeAttr::get(op.getContext(), pipe));

--- a/test/samples/Sync/test_barrier_sync.py
+++ b/test/samples/Sync/test_barrier_sync.py
@@ -1,0 +1,15 @@
+if __name__ == "__main__":
+    # Regression for zhangstevenunity/PTOAS#185:
+    # barrier_sync should support any SyncOpType that can be mapped to a PIPE
+    # (not just TMATMUL/TVEC).
+    print(
+        r"""module {
+  func.func @test_barrier_sync_py() {
+    pto.barrier_sync[<TLOAD>]
+    pto.barrier_sync[<TSTORE_VEC>]
+    pto.barrier_sync[<TVEC>]
+    return
+  }
+}
+"""
+    )

--- a/test/samples/runop.sh
+++ b/test/samples/runop.sh
@@ -291,6 +291,26 @@ process_one_dir() {
       fi
     fi
 
+    # Regression guard for issue #185: barrier_sync must support op types
+    # beyond TMATMUL/TVEC and lower to the expected per-pipe barrier.
+    if [[ "$base" == "test_barrier_sync" ]]; then
+      if ! grep -Fq "pipe_barrier(PIPE_MTE2)" "$cpp"; then
+        echo -e "${A}(${base}.py)\tFAIL\tmissing pipe_barrier(PIPE_MTE2) lowering for barrier_sync[TLOAD]"
+        overall=1
+        continue
+      fi
+      if ! grep -Fq "pipe_barrier(PIPE_MTE3)" "$cpp"; then
+        echo -e "${A}(${base}.py)\tFAIL\tmissing pipe_barrier(PIPE_MTE3) lowering for barrier_sync[TSTORE_VEC]"
+        overall=1
+        continue
+      fi
+      if ! grep -Fq "pipe_barrier(PIPE_V)" "$cpp"; then
+        echo -e "${A}(${base}.py)\tFAIL\tmissing pipe_barrier(PIPE_V) lowering for barrier_sync[TVEC]"
+        overall=1
+        continue
+      fi
+    fi
+
     # Regression guard for issue #117: vector mask must be reset for each
     # `pto.section.vector` region to avoid cross-kernel state leakage.
     # Use an existing sample (Complex/cv_region.py) that contains a vector section.

